src/process.rs

```diff
@@ -1,89 +1,61 @@
+use super::*;
 use colored::Colorize;
-use std::fs;
-use tokio::task;
-
-use crate::memory::ProcessMemoryStats;
-use crate::utils::{can_read_file, get_cmd};
-use crate::AnyError;
-
-pub struct Process {
-    pid: u32,
-    command: String,
-    memory: ProcessMemoryStats,
-}
-
-impl Process {
-    pub fn new(pid: u32) -> Self {
-        let mut process = Self {
-            pid,
-            command: String::new(),
-            memory: ProcessMemoryStats::new(),
-        };
-        process.update().unwrap_or_else(|e| {
-            eprintln!("Error: {}", e);
-        });
-        process
-    }
-
-    pub fn update(&mut self) -> Result<(), AnyError> {
-        self.command = get_cmd(self.pid)?;
-        self.memory.update(&self.pid)?;
-        Ok(())
-    }
-}
+use std::{
+    fmt::{self, Display},
+    fs, io,
+};
 
+#[derive(Default)]
 pub struct Processes {
-    processes: Vec<Process>,
+    mem_stats: Vec<ProcessMemoryStats>,
 }
 
 impl Processes {
-    pub fn new() -> Self {
-        Self { processes: vec![] }
-    }
-
     /// Update the processes
     /// Uses a thread pool to get the memory stats for each process that has a command and can be read from /proc
     ///
     /// # Examples
     /// ```
-    /// let mut processes = Processes::new();
-    /// processes.update()?;
+    /// let mut processes = zmem::Processes::default();
+    /// processes.update().unwrap();
     /// ```
-    pub async fn update(&mut self) -> Result<(), AnyError> {
-        let processes = fs::read_dir("/proc")?
-            .filter_map(|entry| {
-                let entry = entry.ok()?;
-                // Try to parse the file name as a pid
-                let pid = entry.file_name().to_string_lossy().parse::<u32>().ok()?;
-                let command = get_cmd(pid).ok()?;
-                // Only add the process if it has a command and we can read the smaps
-                if command.is_empty() || !can_read_file(&format!("/proc/{}/smaps", pid)) {
-                    return None;
-                }
-                Some(task::spawn(async move { Process::new(pid) }))
-            })
-            .collect::<Vec<_>>();
-
-        // Wait for all the processes to finish
-        let mut processes = futures::future::try_join_all(processes).await?;
-        // Sort the processes by swap usage
-        processes.sort_by_key(|p| p.memory.swap);
-        self.processes = processes;
+    pub fn update(&mut self) -> io::Result<()> {
```

Comment:
Can you please consider the impact on performance while refactoring and find a way to maintain or improve performance without sacrificing functionality? Serializing this function leads to a major decrease in performance. Do not remove concurrency.

Reply:
Operating system does't provide async API for handling file system.

> Then how does async runtime provide async API for file system ? 

It doesn't, tokio basically spawn a OS thread to handle  any blocking operation like filesystem. This is also same for any other async runtime like NodeJS.

From [tokio docs](https://tokio.rs/tokio/tutorial):

> Reading a lot of files. Although it seems like Tokio would be useful for projects that simply need to read a lot of files, Tokio provides no advantage here compared to an ordinary threadpool.

> Can you please consider the impact on performance while refactoring and find a way to maintain or improve performance without sacrificing functionality ?  
 
Initially I did use threads,  But my OS (with kernel v6.2)  didn't like it. With threads it took about 400 ms. without threads it took 440 ms to execute.